posts/2018-08-25-http4s-error-handling-mtl.md

```diff
@@ -0,0 +1,327 @@
+        case UserAlreadyExists(username) => Conflict(username)
+      }
+
+    case req @ PUT -> Root / "users" / username =>
+      req.as[UserUpdateAge].flatMap { userUpdate =>
+        userAlgebra.updateAge(username, userUpdate.age) *> Ok(username)
+      }.handleErrorWith { // compiles without giving you "match non-exhaustive" error
+        case InvalidUserAge(age) => BadRequest(s"Invalid age $age")
+      }
+  }
+
+}
+```
+
+Now we can say this implementation is quite elegant! We are handling and mapping business errors to the according http response and our code compiles without any warning whatsoever. But wait... We are not handling the `UserNotFound` error and the compiler didn't tell us about it! That's not cool and we as functional programmers believe in types because we can know what a function might do just by looking at the types but here it seems we hit the wall.
+
+The problem is that our constraint of type `Sync` from `cats-effect` has a `MonadError` instance with its type error fixed as `Throwable`. So the compiler can't help us here since this type is too generic. And we can't add a constraint for `MonadError[F, UserError]` because we would get an "ambigous implicits" error with two instances of `MonadError` in scope.
+
+So, what can we do about it?
+
+### Next level MTL: Optics
+
+I heard sometime ago about `Classy Optics` (Lenses, Prisms, etc) when I was learning `Haskell` and watched [this amazing talk](https://www.youtube.com/watch?v=GZPup5Iuaqw) by George Wilson but I never got to use this concept in Scala until now!
+
+Well first, let me give you a quick definition of `Lenses` and `Prisms`. In a few words we can define:
+
+- `Lenses` as getters and setters that compose making the accessing of nested data structure's fields quite easy.
+- `Prisms` as first-class pattern matching that let us access branches of an ADT and that also compose.
+
+And the idea of `Classy Optics` is to "associate each type a typeclass full of optics for that type".
+
+***So what am I talking about and how can these concepts help us solving the http error handling problem?***
+
+Remember that I defined the `UserError` ADT by extending `Throwable`?
+
+```scala
+sealed trait UserError extends Throwable
+case class UserAlreadyExists(username: String) extends UserError
+case class UserNotFound(username: String) extends UserError
+case class InvalidUserAge(age: Int) extends UserError
+```
+
+Well there's a reason! By making `UserError` a subtype of `Throwable` we can take advantage of `Prisms` by going back and forth in the types. See what I'm going yet?
+
+`UserRoute` has a `Sync[F]` constraint meaning that we have available a `MonadError[F, Throwable]` instance but we would like to have `MonadError[F, UserError]` instead to leverage the Scala compiler. There's a caveat and is that the error types need to be of the same family so we can derive a `Prism` that can navigate the errors types in one direction or another. But how do we derive it?
```

Comment:
"The caveat is that the..."

Commas for the clause: "constraint, meaning that ... instance, but"